### DISCUSSION TITLE
Why is SQLITE_ENABLE_DBPAGE_VTAB actually needed?

### DISCUSSION THREAD

**Discussion by imciner2** [audacity/audacity#1457] [Development]
I want to unbundle some of the dependencies for the Fedora build, and the new SQLite dependency in 3.0 has a problem. In the Fedora SQLite build we don't have the `SQLITE_ENABLE_DBPAGE_VTAB` build flag enabled, so that isn't available. That then causes the errors like https://github.com/audacity/audacity/issues/784 and https://github.com/audacity/audacity/issues/774. I asked the maintainer of SQLite to enable that flag and they are wondering why exactly Audacity needs this feature (Fedora bug report here: https://bugzilla.redhat.com/show_bug.cgi?id=1973454). They mention that there may be security issues with having it enabled - so they want some more information before they progress.

Why exactly does Audacity need the feature guarded by the `SQLITE_ENABLE_DBPAGE_VTAB` flag? Could Audacity be made to work without that SQLite feature?

**Comment by crsib** [comment]
This extension is used to quickly evaluate the size required by used blocks vs total blocks. Based on that information Audacity decides if there is a need to compact the database on exit. Compacting is not the fastest process, but it is really important due to the way Undo/Redo works. Without it can easily end up with: https://www.reddit.com/r/audacity/comments/p18i1o/86gb_project_file_for_a_12hr_audiobook_recording/

Another important option here is `SQLITE_DEFAULT_PAGE_SIZE=65536`.

We will be working on database compaction and probably the code will be more agnostic of the presence of `sqlite_dbpage`. I'm not sure how quickly we will get there though. 

Also, we will need this extension so we can implement project recovery, similar to sqlite3 `.recover` mode.  We will make the code aware of `sqlite_dbpage`, but in the end, it will mean that recovery won't be available for Fedora users (not that it matters that much, it is an extremely rare case)

**Reply by imciner2** [reply]
Thanks, I will forward that to the maintainers and see if they are willing to enable the option then.

> Another important option here is SQLITE_DEFAULT_PAGE_SIZE=65536

This one is somewhat annoying - since it isn't the default. It does look like it can be overridden inside the Audacity code though - so maybe you could add a `PRAGMA schema.page_size = 65536` database query to the SQLite initialization sequence in the code. That would make the page size correct regardless of the SQLite compile-time options used (https://www.sqlite.org/pragma.html#pragma_page_size).

**Reply by crsib** [reply]
>so maybe you could add a PRAGMA schema.page_size = 65536 database query

We certainly will do it, but it is probably not as easy as it sounds :) #1459 

And, for the sake of consistency, I've opened #1460 as well

**Reply by imciner2** [reply]
Thanks, I have sent the information back to the SQLite maintainers to see what they say. It looks like the page size will be a blocker for me to unbundle it though, since they won't want to change the default page size for the entire distribution.
